### ISSUE TITLE
[core] ProviderScope should offer customizable cache options

### ISSUE THREAD

**Issue by lucavenir** [rrousselGit/riverpod#2353]
**Note**. This issue summarizes this [discord discussion](https://discord.com/channels/765557403865186374/1086701819943723159/1086701819943723159).

**TL;DR.** While we know that scoping isn't really recommended at the moment. But I'd really love to have a compile-safe and hassle-free syntax like this one:
```dart
// caching a scope for just a particular sub-tree
ProviderScope(
  overrides: [...],  // same syntax
  builder: (context, ref, child) {  // optional (or required if we deprecate `child`)
    ref.keepAlive();  // or ref.cacheFor(2.minutes)
    return child;  // lint: don't return something different than `child`
  },
  child: const BookPage(), // required, or deprecated
)
```

### Why this proposal
This allows a scope to have its own caching strategy; here, any scoped provider will live on in its own scope for the next two minutes. No need to formulate weird family shenanigans to implement a relatively simple caching strategy.

Specifically, any provider (directly / indirectly) using `toBeOverriddenProvider` should just specify their `dependencies`, without having the need to write down a coupled `.family` provider just to cache its values. And that's it! This subtree (or feature) can really easily be maintained.

### What's wrong with `.family`?
**Note**. A full reproducible examples can be found on [here](https://github.com/lucavenir/family_or_scope/).

There's nothing wrong with family, really. But take the repo above. We want to be able to navigate between two routes, a home route and a sub slug route, and cache their results.
The home route shows a simple list with clickable tiles. Each tile routes towards its associated `:id` page, which shows results from a parameterized API. Caches should stay alive for, say, 2 minutes.
```log
Full paths for routes:
  => /
  =>   /:id
```
### Solution 1
**TL;DR.** This solution is shown in the `main` branch.

To solve this, we propose the use of a _to-be-overridden_ provider.
```dart
@Riverpod(dependencies: [])  // the to-be-scoped provider
int appleId(AppleIdRef ref) => throw UnimplementedError('This provider is meant to be overridden.'); 

@riverpod  // cache the apple instance
Apple apple(AppleRef ref, int id) {
  final apples = ref.watch(applesProvider);  // the list of elements
  final apple = apples.firstWhere((element) => element.id == id);

  ref.cacheFor(2.minutes);
  return apple;
}

@riverpod  // cache the info API call
Future<int> appleInfo(AppleInfoRef ref, int id) async {
  final apple = ref.watch(appleProvider(id));
  final apiResult = Random().nextInt(10);

  final result = Future.delayed(800.milliseconds, () => apple.id + apiResult);

  ref.cacheFor(2.minutes);
  return result;
}
```
This requires the UI to `watch` our `appleIdProvider` first, and then pass its result to the families to make 'em work.

**The good news**: This solution works.
**The bad news**: Code duplication.

### A mistake
We don't like retrieving the `id` each time. We like our code DRY or at least a little bit WET. Thus, we modify `appleProvider`:
```dart
// WARN: pseudocode
@Riverpod(dependencies: [appleId])
Apple apple(SelectedAppleRef ref) {
  final id = ref.watch(appleIdProvider);
  final apples = ref.watch(applesProvider);
  final apple = apples.firstWhere((element) => element.id == id);

  ref.keepAlive();
  return apple;  
}
```
Ah! This is simpler, right? We even gave up on families and .._riiiight_, this obviously won't work. We _need_ the families or else our caching breaks.
What's frustrating about this is that this provider _looks_ like it should work. Except, well, it doesn't: `appleId` gets re-scoped everytime we navigate there. Thus, any cache will be discarded when leaving that scope, since scoped providers get disposed.

### Solution 2
**TL;DR.** This solution (or rather, the current state-of-art / current workaround) is shown in the `current-apple` branch.

We can't really afford to give up on cutting down code duplication: it's actually really easy to have dozens of `ref.watch` calls, just to retrieve `id`, needed to inject it into its family.

Thus, for each "familied" provider, we _must_ construct its own scoped provider as follows:
```dart
@riverpod  // for each one of these...
Future<int> appleInfo(AppleInfoRef ref, Apple apple) async {
  final apiResult = Random().nextInt(10);

  final result = Future.delayed(800.milliseconds, () => apple.id + apiResult);

  ref.keepAlive();
  return result;
}

@Riverpod(dependencies: [currentApple])  // ... we need this
FutureOr<int> currentAppleInfo(CurrentAppleInfoRef ref) {
  final currentApple = ref.watch(currentAppleProvider);
  return ref.watch(appleInfoProvider(currentApple).future);
}
```
This works fine, even though it's kind-of tedious. It respects a WET approach and I'm fine with it.

### Why this DX is bad

While it is not still clear if scoping a subtree should be done at all to solve this, we encountered the following non trivial (_at all_) problems:
1. Scoping isn't easy. It comes at the cost of specifying `dependencies` and it's really easy to lose track of the dependency chain. Without the linter, it simply would be impossible
2. Scoping - as of now - heavily influences the rest of the subtree in a way that's often unexpected and inconsistent as compared to of the rest of the framework
3. Maintaining this code is hard; this feels like a large caveat to me. It should be documented why I didn't just `watch` the `id` directly in my `apple` providers. It also be mentioned that this shouldn't be done in the future, ever, when wanting data to be cached for this single `apple`
4. Families don't entirely solve the code duplication issue. As mentioned, for each `apple` family provider, we need a duplicated `currentApple` provider. In my own real-life scenario, I have ~15ish to-be-familied providers with their own `current` duplication... just because I want some cache (which is Riverpod's responsibility)
6. This is a gut feeling more than anything, but I honestly felt like families are a workaround rather than "the right way" to do it. [Families are meant for a different usage](https://docs-v2.riverpod.dev/docs/concepts/modifiers/family) , whereas here they are being used to work around a scoping issue. `.family` feel like a hammer, and when all you have is a hammer...

**Additional context**
I tried so many different options, as shown in the discussion. I'm open to any suggestion at this point.

**Comment by rrousselGit** [comment]
> TL;DR. Scoping providers either puts at risk the ability to offer highly customizable caching strategies or just badly hurts code maintainability. We shouldn't think about this tradeoff, at all.

Why?


> We even gave up on families and ..riiiight, this obviously won't work. We need the families or else our caching breaks.

Why?


> Thus, any cache will be discarded when leaving that scope, since scoped providers get disposed.

You mean when destroying ProviderScope? Sounds logical right?
What would happen in your scenario then? Because if the ProviderScope is destroyed, any for of cacheTime would have no impact.

**Comment by rrousselGit** [comment]
> Families don't entirely solve the code duplication issue. As mentioned, for each apple family provider, we need a duplicated currentApple provider. In my own real-life scenario, I have ~15ish to-be-familied providers with their own current duplication... just because I want some cache (which is Riverpod's responsibility)


That's a non-issue.

I don't think having to pass the ID around is a problem. It's very much fine in fact.
Using scoping to work around that is the issue I'd say. I discourage doing it.

**Comment by rrousselGit** [comment]
If your issue is indeed with unmounting/remounting ProviderScope, this has nothing to do with families.
The problem is scoping. And scoping has always been considered a discouraged feature.

But I don't see anything wrong with what you've described.


Chances are you misunderstood the implications of scoping. You scoped the provider. Therefore it's obvious that you'll get one state per scope.

So doing something like:

```dart
@Riverpod(dependencies: [id]);
Future<T> fetchT(FetchTRef ref) {
  final id = ref.watch(idProvider);
  return fetch('/api/$id');
}
```

is obviously a bad idea if you planned to scope that ID yet share the state outside said scope.




Just don't use scoping. Pass the params around. Nothing wrong with that. A bit tedious sometimes, sure. But it all works.

**Comment by davidmartos96** [comment]
I agree with OP.
Families are great, but on big applications with complex features tied to a specific entity (say the apple), one would want to create multiple providers to split the state. To problem is that then you probably need to convert everything into a family, which affects readability and makes it harder to create new providers.
I also get the feeling that I'm using riverpod wrong sometimes when it comes with complex state + families/scope.
I would like a way to reason about a certain part of the application in an isolated way with riverpod data flow/caching which is where it shines, without the need to turn everything into a family or pass an id everywhere. I'm talking about complex features where there would be many widgets and state involved. Having to pass it around adds quite a lot of friction to Flutter's way of splitting widgets.

Just wanted to share because I felt very much identified with what was discussed in the discord thread.

**Comment by rrousselGit** [comment]
The thought about #2356 is that we could then (and only then) rely a bit more on scoping.

For example one of @TimWhiting's early proposals was to have families have a "default". To enable both:

```dart
ref.watch(family(42));
ref.watch(family.default);
```

where we can do:

```dart
ProviderScope(overrides: [family.default.overrideWith(...)])
```

I didn't want to promote this because I didn't want to promote scoping. But if we can make scoping compile-safe through lints, that might change the game.
Although I still think it should be used with parsimony. But for that use case, it could be fine.

**Comment by lucavenir** [comment]
Thanks for the feedback here and on discord.

I really appreciate your involvement with this. I'm investing a little bit more because I truly want a "go-to" reference for the community having the problem. I'm pretty sure that a `/feature` + `/feature/:id` is a pretty common scenario for any application.

I always to promote Riverpod (with clients and colleagues) as being easy and scalable, which it totally is, once I also manage to make sense out of this non trivial detail.

> I don't see anything wrong with what you've described.
> Why? [on TLDR of this issue]
> Why? [on families being needed]
> I don't think having to pass the ID around is a problem. It's very much fine in fact.

I think I answered here and on discord, but let's summarize for the sake of the readers:
1. I want to cache multiple `apples`: to do so I need `.family`, or else I could only cache only _one_ `apple` at a time
2. In order to use `.family`, I must pass `id`s around. But this is **not** scalable in a very large feature: consider `/:id` having a rich UI experience (e.g. tabs) and maybe even more sub pages or sub slugs (e.g. `/apples/:appleId/farmers/:farmerId`). We would need to pass even more `id`s around. This quickly becomes complicated.
3. The **only** reliable workaround I've found is shown above. **For _each_ variable that needs caching**, we should (1) Define a `.family` provider (e.g. `apple(int id)`) and (2) Define a `currentApple` provider that watches a scoped `appleId` provider (we must specify dependencies). Whenever we need to inject `apple` in our `/:appleId` page, we can `watch` our `currentApple` and let `apple(int id)` do the caching for us.

You _do_ see why this is at least kind-of inconvenient, right? Is there anything I'm not seeing? Any alternative?

> ... destroying ProviderScope? Sounds logical right?

Yep, that's the point of this proposal. No one wants to change this logic, which is important and it's there to stay. I want to enhance it, giving the possibility to implement caching strategies for a scope, so that it doesn't _immediately_ get disposed.
At the end of the day, it would be a very similar experience to the `.family` modifier, but with a _very_ simplified developer experience. No caveats, no paper cuts, no need to be _very careful_ about scoping. No ids being passed around. You'd get the same exact APIs we use for the rest of the framework.

> Using scoping to work around that is the issue I'd say. I discourage doing it.
> And scoping has always been considered a discouraged feature.

Just to be clear I _never_ used scoping nor overrides ever before, besides testing of course. I'm not searching a solution with `ProviderScope` because I love it. I'm trying not to duplicate code all around and scopes just appeared before me.
I often see that you're skeptic about scoping providers (which is funny since you're the creator of this magnificent ecosystem 😹). So, in your opinion, should overriding providers be avoided _in general_? Except testing, of course.

> [scoping] is obviously a bad idea if you planned to scope that ID yet share the state outside said scope.

No, you got this wrong. No one wants to use `id` outside the scope. The point is that we have a slug, i.e. a parametrized page, which can be very much complex. In contrast, I'm being told to pass around `id` for the whole subtree: unfeasible.
The thing is, the more I speak about scoping, the more it feels it _should_ be the right solution. That sub-tree indeed should use a scoped value that must _not_ be used outside. Adding some cache API to it is just a little beneficial prop (coherent with providers).

> I don't see anything wrong with that [using families all around]. That's the end-game of Riverpod yes, and it's great.

No, I'm sorry, but no, it isn't, and I've shown here (and on discord / my repo) why. It's not scalable thus feasible for non-trivial applications. As I said above, it feels a "when all you have is a hammer, everything looks like a nail" situation.

> The problems typically occur when folks try to not to use them or hide them [families]. Like what's described here or folks who use a StateProvider for the "current ID"
> Sure, there is a bit of friction with having to pass IDs around. But that's fairly minor.

Again, I disagree. I don't really care about scoping - it's not like I love scoping or anything. I never, ever, used it before.
The funny thing is: I walked this path **exactly because I am trying to _not_ use `StateProvider` to hold this `id`**. I find this funny because I am _really, really_ trying to use `.family` all around, but it turns out to be a hell of a dev-experience.

> Here's a thought on how to make scoping better https://github.com/rrousselGit/riverpod/issues/2356.

At the moment I can't really see see how this lint, nor a `.family.default`, would help this issue.

### The bottom line

In the end, the proposal is to let scopes have their own caching strategy to enable a smooth DX without forcing everything into `.family`. Why are you against it in the first place?

I don't exactly want scopes to replace families or anything like that. This issue can be summarized as: we _like_ family providers, but overusing them forces us to write fundamentally duplicated code everywhere. If a "scope cache strategy" is not possible, then I'm all ears for alternatives, which allow me to write SOLID code. I'm not a DRY nazi, but this is absurd.

**Comment by rrousselGit** [comment]
I really appreciate all the effort you're putting into this. But I believe there's some communication issue here.   
You mix-and-match too many things at once in this issue

It's a testament to how passionate you are about Riverpod and shows how frustrated you are with the issue. But it also makes the discussion a bit difficult because there are too many things to unwrap.

____


Ultimately these are the following facts about this discussion:

- `family` has a tendency to cause props drilling (having to pass parameters to all widgets)
- Officially, I recommend not fighting against the props drilling. It's inconvenient for sure, but a small price to pay. But you disagree with this.
- Les officially, you can use scoping to inject in the widget tree the family arguments.
- Scoping is considered "advanced" and generally discouraged because not compile-safe.
- Even though scoping is a solution to the family props drilling, because of the above point, Riverpod
  makes the scoping solution more difficult than it has to be because it's discouraged.
- The scoping issue could likely be solved through riverpod_lint, making scoping statically safe through lints.
- If this is achieved, then Riverpod could change its stance on scoping and promote it a bit more.
  One use-case is the inconvenience you're facing: Fixing props drilling with `family`


There have been many discussions about the props drilling of families already.

There also have been proposals to improve this issue, such as the one I mentioned earlier: https://github.com/rrousselGit/riverpod/issues/767


___ 

Last but not least, your proposal is unclear to me. Due to this issue being a bit all-over the place, there's very little information related to your proposal in the end.

I do not understand how `ref.addOverride` is different from that `overrides: []` parameters, nor do I understand how it works.

Would it be possible to make an issue hyper-focused on that? Where we could debate on the proposal without fighting on what the problem is.
We clearly agree that there are some inconveniences (although don't agree on their importance). So if you have a solution, I'm all ears. But so far, I don't know what this does and how it'd work/

**Comment by rrousselGit** [comment]
I don't think sharing the ProviderContainer is reasonable. Because then it means that a ProviderContainer can have two different parents, because the same ID can be alive in two different widgets. The route `/books/42` could be present twice in the navigation history for instance.

It _might_ be doable, but it's such a large core change that it's hard to see all the possible consequences.
I also don't like the idea of introducing a random "containerID" parameter. It's typically the sort of thing I wanted to avoid with Riverpod.

I'm not sure how this would be preferable to https://github.com/rrousselGit/riverpod/issues/767, which is simply syntax sugar over the existing scoping mechanism.

**Comment by rrousselGit** [comment]
I don't see a way for this to be possible without sharing the container between multiple widgets. You'd need some form of ID, which means the container would be shared between routes.

But again, why would this be preferable over https://github.com/rrousselGit/riverpod/issues/767?

**Comment by AhmedLSayed9** [comment]
I think the proposed solution by Tim will solve the code duplication drawback of `Solution 1`, so to clarify I got it right instead of:
```dart
    final id = ref.watch(appleIdProvider);
    final info = ref.watch(appleInfoProvider(id));
```
It'll be possible to do:
```dart
    final info = ref.watch(appleInfoProvider.scoped);
```

but tbh I find that syntax is hard to understand. like first question will come in mind why not using a scoped `appleInfoProvider` instead of `appleInfoProvider.scoped` & the answer will be: `appleInfoProvider.scoped` is not in fact a scoped provider but it's a normal family provider that uses a scoped provider, because we want it to be long-lived when we leave the scope. I'd suggest this instead:
```dart
  //A non-scoped provider that uses scoped provider
  @Riverpod(dependencies: [currentApple], scoped: false)
  Future<int> appleInfo(AppleInfoRef ref) {
    final apple = ref.watch(currentAppleProvider);
    final random = Random().nextInt(10);
  
    return Future.delayed(500.milliseconds, () => apple.id + random);
  }

  //Clearly inside build
   final info = ref.watch(appleInfoProvider);
```

**Comment by rrousselGit** [comment]
Don't focus too much on the ".scoped" name.  We can pick a different name & co. It's more about the idea as a whole.


The point was mainly to not use the provider itself (as it avoid `ref.watch(family)` and use either `ref.watch(family(...)` or `ref.watch(family.something)`).  
The goal is that this makes refactoring smooth. Refactoring a non-family to a family would silently failed if we used `ref.watch(family)`) whereas it'd cause a compilation error with this syntax.     
So in that sense we'd do the same thing with your suggestion.

____


Your proposal is not possible. A non-scoped provider cannot use a scoped provider. This breaks the unidirectional dataflow of providers

**Comment by AhmedLSayed9** [comment]
You're right about my proposal, the scoped container will have to get back to the root one to store a provider in there.

So, the suggestion with a family provider that takes in a provider as a param, that param will update on changes right?

so it's like imaginatively doing:
```dart
  Future<int> appleInfo(AppleInfoRef ref, ref.watch(currentAppleProvider) apple) {
    final random = Random().nextInt(10);
    return Future.delayed(500.milliseconds, () => apple.id + random);
  }
```
in addition that currentAppleProvider can be scoped

**Comment by rrousselGit** [comment]
ref.watch and familiy parameters don't quite behave the same, but yes, pretty much.    
What you did is effectively the same as https://github.com/rrousselGit/riverpod/issues/767 but on a per-parameter basis instead of on the whole argument list

**Comment by lucavenir** [comment]
> I really appreciate all the effort you're putting into this

I actually appreciate _your_ effort replying into this thread; it shows you're unbiased and that you actually care about the community and this ecosystem.

> Officially, I recommend not fighting against the props drilling. It's inconvenient for sure, but a small price to pay. But you disagree with this.

This is the only statement that really threw me off for a second.

Prop drilling is _the_ problem for any component-based framework and - for me - it is the main reason I'm looking for a state management library such as Rivrepod. I stumbled upon this ecosystem (and I love it!) exactly to "fight" this.

Being able to choose dep injection instead of prop drilling is not just some syntactic taste. It is not because we don't like writing a few more characters. Instead, I'm pretty sure it is widely accepted that prop drilling doesn't help maintainability.

Looking back at this I say: I'm not even sure why one would look at a state management library if the state management library itself reccomends prop drilling for slug parameters / pages / views / other use case, etc.

> Scoping is considered "advanced" and generally discouraged because not compile-safe.

I totally understand why you would not recommend this. But, without doing it on purpose, I found myself loving scopes because they're way simpler than they look and - when using them - it looks like they solve exactly the problem above.

Isn't there _really_ any solution to make scoping compile-time safe?

### About other proposals
Initially, I didn't quite understand #767. Now, I see that the proposal is ultimately about the following syntax sugar:

```dart
// Desugars to ref.watch(myFamilyProvider(ref.watch(myFamilyProvider.paramProvider)));
ref.watch(myFamilyProvider.scope);
```

This would solve the prop drilling issue for _one_ parameter.

What about routes such as `authors/42/books/42/citations/42`? There, we would need _three_ scoped parameters, one for each subroute. My concern is: is it possible to desugar these via proposal #767? If yes, we could definetively re-open that proposal. But before closing this one...

### About this proposal
Yes, I ultimately was proposing to implement a cache mechanism onto `ProviderScope`. I thought of this to make the use of scopes even easier and ultimately recommended. I realize this proposal kind-of goes "against" your vision about Riverpod as of now.

My concern is that proposal #767 isn't exactly easy to understand even for a seasoned Riverpod user. It involves a misterious desugaring syntax of some sort.

Given it is feasible, I think this proposal is simpler: it would let the developer implement their caching strategy onto the scoped parameter, so that Riverpod doesn't drop it immediately when exiting the scope. This is something we already do with Providers. It is familiar and safe.

This would allow `/book/42` and `/book/43` to cache their providers without relying on families, props, prop drilling and fancy-footwork-syntax-sugar. Caching and handling state onto complex paths like `authors/42/books/42/citations/42` would require zero overheads with a familiar syntax.

I realize that `addOverride` threw this thread off. I'm sorry, discard that. Ultimately I would love `ProviderScope` to be used like this:

```dart
// caching a scope for just a particular sub-tree
ProviderScope(
  overrides: [...],  // same syntax
  builder: (context, ref, child) {  // optional (or required if we deprecate `child`)
    ref.cacheFor(2.minutes);  // This `ScopeRef` could be restricted to use just `keepAlive()`
    return child;  // lint: don't return something different than `child`
  },
  child: const BookPage(), // required, or deprecated
)
```

Again - I'm not sure if this is feasible. I did not understand the "sharing a container" discussion above.

Here, I just wanted to show what I meant with that weird `addOverride` syntax. I realize I got way too much imaginative that day 😆

**Comment by rrousselGit** [comment]
> Instead, I'm pretty sure it is widely accepted that prop drilling doesn't help maintainability.

I disagree. I believe it's a case of personal bias. IMO the problem is minor, but the inconvenience makes it appear to be worse than it is.  

My stance on props drilling is that, in the context of widgets and their declarative nature combined with Dart's type-safety, props drilling isn't bug-inducing.   
It's mostly a cause of boilerplate and makes refactoring tedious. But it's not a huge deal. Everybody knows how to pass parameters to widgets.


On the flip side, trying to solve props drilling has huge dangers. Realistically, any solution is likely to have many pitfalls; could decrease readability (by making it harder to follow where the props are going/coming from); and could increase complexity (new APIs, new things to learn).

So the dangers are huge, but the benefits aren't that big (typing a bit less on the keyboard).


In particular,  InheritedWidgets are one of the hardest concepts to grasp in Flutter.


> Prop drilling is the problem for any component-based framework and - for me - it is the main reason I'm looking for a state management library such as Rivrepod. I stumbled upon this ecosystem (and I love it!) exactly to "fight" this.

_For you_ :D

There are many problems way more important than props drilling.    
Riverpod significantly reduces the likeliness of errors & improves composability through its delcarative/functional programming with ref.watch & AsyncValue + other. 

Props drilling isn't hard. It's tedious, but anyone can do it. You don't need to be a senior to know how to pass parameters to widgets.    
Having chainable FutureProviders which automatically recompute whenever their dependencies update, with automatic error handling, all while being testable? That's a lot harder to do.

> But, without doing it on purpose, I found myself loving scopes because they're way simpler than they look and - when using them - it looks like they solve exactly the problem above.

Way harder than they look you mean :P

Keep in mind, people have a **very** hard time understanding InheritedWidgets. Tons of issues/stackoverflow questions are based on a misunderstanding of InheritedWidgets.   

From my experience, they cause way more problems than they solve in the context of business logic.

> Isn't there really any solution to make scoping compile-time safe?

There's an issue with this which I'm working on https://github.com/rrousselGit/riverpod/issues/2356.  
That's part of the motivation behind riverpod_lint.

> Initially, I didn't quite understand #767. Now, I see that the proposal is ultimately about the following syntax sugar:
> 
> ```dart
> // Desugars to ref.watch(myFamilyProvider(ref.watch(myFamilyProvider.paramProvider)));
> ref.watch(myFamilyProvider.scope);
> ```
> 
> This would solve the prop drilling issue for _one_ parameter.

Not for "one parameter" but "one provider".  
A provider could have 10 parameters that you'd still do `ref.watch(myFamilyProvider.scope);`

What'd change is the override line

```dart
@riverpod
Model fn(FnRef ref, {required int userId, required int bookId}) {...}

ProviderScope(
  overrides: [
    fnProvider.withDefault(userId: 42, bookId: 21),
  ],
)
```


> My concern is that proposal https://github.com/rrousselGit/riverpod/issues/767 isn't exactly easy to understand even for a seasoned Riverpod user. It involves a misterious desugaring syntax of some sort.

I don't think so.

I'm pretty sure #767 is more what you're asking that what you're proposing. But chances are #767 isn't described quite well. The issue is pretty old too. Codegen wasn't a thing by then, nor was `dependencies` (which is what closed the issue ;) )


> Given it is feasible, I think this proposal is simpler: it would let the developer implement their caching strategy onto the scoped parameter, so that Riverpod doesn't drop it immediately when exiting the scope. 


As I've tried to argue before, I don't think it's feasible. Nor do I think it's the correct abstraction.  
You're suggesting to cache the cache.  That's not the correct place IMO.    

The solution should be at the provider level instead.  
Trying to do it at the ProviderScope level would likely have too many fundamental problems.

For example, one scenario that would be hard to support with your proposal is to support having two different routes point to the same state.  

Say we have `/users/:userId/comments/:commentId` and `/threads/:threadId/comments/:commentId`:  
If the comment ID is the same, the stake of the "commentProvider(id)" should be the same.  
That'd naturally happen with `family`, and would stay this way with #767. But with your proposal, this could break.

And that's just an example on top of my head. I'm sure we could come up with many more.    
Hence why I keep arguing that scoping is hard and that props drilling objectively ain't that bad in comparison. So many things could go wrong with scoping.

**Comment by TimWhiting** [comment]
```dart
@riverpod
Model fn(FnRef ref, {required int userId, required int bookId}) {...}

ProviderScope(
  overrides: [
    fnProvider.withDefault(userId: 42, bookId: 21),
  ],
)```

This looks great! Add in a lint to make sure that any required parameters have some withDefault override in a provider scope and this would be a perfect solution I think. (The lint could get tricky with multiple provider scopes & pieces of the application which don't need or use those providers, but I think it is doable).

**Comment by rrousselGit** [comment]
A lint to make scoped providers compile-safe & handle overrides is definitely doable.   
I've just started working on this and am currently investigating how realistic it is. But I'm pretty optimistic.

This will most definitely require a lint for warning against dynamic provider invocations such as:
```dart
class MyWidget {
  final ProviderListenable<T> provider;

  build(context, ref)  {
    // Bad, dynamic provider invocation
    ref.watch(provider);
  }
}
```

As such code makes it very difficult to know what `MyWidget` is using and therefore whether it's safely accessing scoped providers or not.

That's one of the reasons I've always pushed back against dynamic provider invocations. Keeping everything static empowers `riverpod_lint` to do a lot

**Comment by lucavenir** [comment]
Hi, I just wanted to thank Remi again for the effort on this issue.

For my own sake, I could say this discussion was very much worth it. I hope this issue helped the community too, by letting Remi explain the "right direction" on the solution.

After these last few messages I'm 100% convinced that the direction is #767, with the right syntax and lints to make it compile time safe.

I'll be honest though, I didn't understand #2356. I don't get what `@Scopes` is, why it is applied to a `Widget`, why it requires overrides to be statically analyzable and what code is going to produce.

> But chances are #767 isn't described quite well. The issue is pretty old too.

We could move this discussion there and find the best syntax, even though the last messages here show a great API imho. Maybe we should add a `TLDR` there with syntax to declare a family with default params and how to invoke the default family.

Thank you again for your time! 🙏🏽🙏🏽🙏🏽

**Comment by rrousselGit** [comment]
> I didn't understand https://github.com/rrousselGit/riverpod/issues/2356. I don't get what @Scopes is, why it is applied to a Widget

It's an annotation. It does nothing besides the analyzer to understand your code.

Technically the code is analyzable without those annotations, but the problem is efficiency. The annotation makes the analysis waayyy faster.  

The annotation is self-managed, like `dependencies` on providers when using the `provider_dependencies` lint. There's a quick fix to automatically add/update/remove the annotation depending on what you use inside the widget.  
So you wouldn't actually write it yourself
